util/sock.c

```diff
@@ -297,21 +298,38 @@ sock_resolve(const char * addr)
 }
 
 /**
- * sock_resolve_one(addr):
+ * sock_resolve_one(addr, addport):
  * Return a single sock_addr structure, or NULL if there are no addresses.
  * Warn if there is more than one address, and return the first one.
+ * If ${addport} is non-zero, use sock_addr_ensure_port() to add a port number
+ * of ":0" if appropriate.
  */
 struct sock_addr *
-sock_resolve_one(const char * addr)
+sock_resolve_one(const char * addr, int addport)
 {
 	struct sock_addr ** sas;
 	struct sock_addr * sa;
 	struct sock_addr ** sa_tmp;
-
-	/* Resolve target address. */
-	if ((sas = sock_resolve(addr)) == NULL) {
-		warnp("Error resolving socket address: %s", addr);
-		goto err0;
+	char * bind_addr;
+
+	/* Resolve target address... */
+	if (addport == 0) {
+		/* ... using the original address. */
+		if ((sas = sock_resolve(addr)) == NULL) {
+			warnp("Error resolving socket address: %s", addr);
+			goto err0;
+		}
+	} else {
+		/* ... after ensuring that it has a port (if appropriate). */
+		if ((bind_addr = sock_addr_ensure_port(addr)) == NULL) {
+			warnp("sock_addr_ensure_port");
+			goto err0;
+		}
+		if ((sas = sock_resolve(bind_addr)) == NULL) {
+			warnp("Error resolving socket address: %s", bind_addr);
+			goto err0;
+		}
+		free(bind_addr);
```

Comment:
This seemed like the most "self-contained" implementation of using `sock_addr_ensure_port()`, but as a result, it will report problems (below) using the original address, rather than the version containing `:0`.

I've gone back and forth about this a few times.  Sometimes I think that's desirable (since the original address is what people would have given it on the command line), sometimes I think it's sloppy (and we should be reporting the `bind_addr`).

On a related note: if you think it's ok to report problems with `addr`, then I should do the same thing for the above warning.